### PR TITLE
[RHCLOUD-24236] manage user rights on template api

### PR DESCRIPTION
**Pull request by g-duval**
Implementing user rights checks with following rules:
- A common template don't need to belong to any application. Only internal admins can manage them.
- A template linked to an application can only be created/updated/deleted by an internal user with this application admin role.

- An instant template can use any templates of any application
- An instant template without event type can be created/updated/deleted by any internal user
- An instant template with event type can only be created/updated/deleted by an internal user with this application admin role.

- An aggregation template can use any templates of any application
- An aggregation template without application can be created/updated/deleted by any internal user
- An aggregation template with an application can only be created/updated/deleted by an internal user with this application admin role.

### CODE REVIEW

**Review by MikelAlejoBR — APPROVED** [review]
LGTM! Nice!

File: backend/src/main/java/com/redhat/cloud/notifications/routers/internal/TemplateResource.java [hunk 1]
```diff
@@ -56,8 +65,12 @@ public class TemplateResource {
...
     @Produces(APPLICATION_JSON)
     @Transactional
-    @RolesAllowed(RBAC_INTERNAL_ADMIN)
-    public Template createTemplate(@NotNull @Valid Template template) {
+    @RolesAllowed(RBAC_INTERNAL_USER)
+    public Template createTemplate(@Context SecurityContext sec, @NotNull @Valid Template template) {
+        // Common templates don't have any applicationId, only internal admins can manage them
+        if (!sec.isUserInRole(ConsoleIdentityProvider.RBAC_INTERNAL_ADMIN)) {
+            securityContextUtil.hasPermissionForApplication(sec, template.getApplicationId());
+        }
```
[the remark this answers is not in the document]
Comment (g-duval) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/internal/TemplateResource.java]: Right, thanks

File: database/src/main/resources/db/migration/V1.73.0__RHCLOUD-24236_add_application_fk_to_templates.sql [hunk 2]
```diff
@@ -0,0 +1,3 @@
+ALTER TABLE Template
```
[the remark this answers is not in the document]
Comment (g-duval) [reply database/src/main/resources/db/migration/V1.73.0__RHCLOUD-24236_add_application_fk_to_templates.sql]: Right, thanks

**Comment by gwenneg** [comment]
@g-duval Could you please rebase this PR?

File: common/src/main/java/com/redhat/cloud/notifications/models/Template.java [hunk 3]
```diff
@@ -33,6 +36,13 @@ public class Template extends CreationUpdateTimestamped {
     @NotNull
     private String data;
 
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(name = "application_id")
+    private Application application;
```
[the remark this answers is not in the document]
Comment (g-duval) [reply common/src/main/java/com/redhat/cloud/notifications/models/Template.java:43]: Sure

**Comment by gwenneg** [comment]
I'm super late to the party, sorry about that :disappointed: 

> * A common template don't need to belong to any application. Only internal admins can manage them.

This makes sense. However, I wonder if we should identify these common templates with a boolean flag in the DB rather than the absence of `applicationId` field. If someone creates a template without selecting an app at first and then realizes it was a mistake, they won't be able to fix that mistake without our help. Allowing to update a Template that is not `common` (DB flag = false) and link it with an app would make sense to me.

If the flag was introduced, we would need to reject common template (flag = true) with an `applicationId` field value.

> * A template linked to an application can only be created/updated/deleted by an internal user with this application admin role.

:+1: 

> * An instant template can use any templates of any application

This sounds very risky to me. There's a limitation with the old templates design that may make it mandatory (for now): the only way to differentiate a draft instant template from a finished one is the association of the instant template with an event type. As soon as an instant template is associated with an event type, it becomes available when events are processed and emails are sent. As a consequence, creating a template without a link to an event type makes sense but this becomes an issue now that we want to introduce permissions... I think we should make the link between the instant template and an event type mandatory, and rely on a new DB boolean flag to identify drafts. We would only allow one "non-draft" template linked to an event type at a given time.

> * An instant template without event type can be created/updated/deleted by any internal user

As I wrote above, I wouldn't allow this anymore.

> * An instant template with event type can only be created/updated/deleted by an internal user with this application admin role.

:+1: 

> * An aggregation template can use any templates of any application

This again sounds risky to me. I would address this concern with the same kind of solution than for instant templates: a mandatory link between the aggregation template and an application and a new DB boolean flag to identify drafts.

> * An aggregation template without application can be created/updated/deleted by any internal user

I wouldn't allow this anymore.

> * An aggregation template with an application can only be created/updated/deleted by an internal user with this application admin role.

:+1: 

I'm really interested in your opinion (everyone) about this so don't hesitate to comment what I wrote and suggest a different approach! :smile:

**Comment by gwenneg** [comment]
cc @josejulio

**Comment by josejulio** [comment]
> This makes sense. However, I wonder if we should identify these common templates with a boolean flag in the DB rather than the absence of applicationId field. If someone creates a template without selecting an app at first and then realizes it was a mistake, they won't be able to fix that mistake without our help. Allowing to update a Template that is not common (DB flag = false) and link it with an app would make sense to me.

What other ability would the separate flag give us? 
We could refuse to save a template in the first place if a non admin is saving without an application.

> > An instant template can use any templates of any application

I think templates can only use common templates or their own templates to prevent from stepping on each other toes. It's not the same to depend on stuff that we provide, than depend on stuff that other apps provide - it will be harder to coordinate.

> I think we should make the link between the instant template and an event type mandatory, and rely on a new DB boolean flag to identify drafts.

The only issue i see here is that we won't allow to share templates within an application. i.e.
`headers_of_tables` being used by `event-a`, `event-b`, `event-c`; where only the later are linked to an event type.

> We would only allow one "non-draft" template linked to an event type at a given time

What would happen if we want to make changes to a non-draft template? I suppose we would need to duplicate the template, make changes and promote it to "non-draft" once that's ready? (we need to agree on a name here :joy:) and this is where versions will kick in, right?

I think we need to review and further discuss how we are going to proceed here - all the heavy lifting is going to happen on stage -but will need to replicate it to prod. We will end up with 2 different "histories" if we allow the users to handle that manually - and that might be confusing (and error prone). We need to analyze the sync mechanism. 

I think it's better if we move the discussion to a document - it's hard to keep track of the different topics in a single thread :sweat_smile:

**Comment by g-duval** [comment]
feature cancelled.

**Pull request closed without merging**